StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentMethodType.swift

```diff
@@ -229,52 +229,141 @@ extension PaymentSheet {
             return paymentTypes
         }
 
+        /// Returns whether or not PaymentSheet, with the given `PaymentMethodRequirementProvider`s, should make the given `paymentMethod` available to add.
+        /// Note: This doesn't affect the availability of saved PMs.
+        /// - Parameters:
+        ///   - paymentMethod: the `STPPaymentMethodType` in question
+        ///   - requirementProviders: a list of [PaymentMethodRequirementProvider] who satisfy payment requirements
+        ///   - intent: a intent object
+        ///   - supportedPaymentMethods: the payment methods that PaymentSheet can display UI for
+        /// - Returns: true if `paymentMethod` should be available in the PaymentSheet, false otherwise
         static func supportsAdding(
             paymentMethod: PaymentMethodType,
             configuration: PaymentSheet.Configuration,
             intent: Intent,
             supportedPaymentMethods: [STPPaymentMethodType] = PaymentSheet.supportedPaymentMethods
         ) -> Bool {
-            if let stpPaymentMethodType = paymentMethod.stpPaymentMethodType {
-                return PaymentSheet.supportsAdding(
-                    paymentMethod: stpPaymentMethodType,
-                    configuration: configuration,
-                    intent: intent,
-                    supportedPaymentMethods: supportedPaymentMethods
-                )
-            } else if case .dynamic = paymentMethod {
-                return supports(
-                    requirements: paymentMethod.supportsAddingRequirements(),
+            guard let stpPaymentMethodType = paymentMethod.stpPaymentMethodType else {
+                if case .dynamic = paymentMethod {
+                    return supports(
+                        requirements: paymentMethod.supportsAddingRequirements(),
+                        configuration: configuration,
+                        intent: intent
+                    )
+                }
+
+                return false
+            }
+
+            switch intent {
+            case .paymentIntent(let paymentIntent):
+                // if setting up, fall through to the setup case
+                if paymentIntent.setupFutureUsage != .none {
+                    fallthrough
```

Comment:
Between the nested `if` inside `guard` above, the `fallthrough`, I'm finding the control flow pretty hard to understand.

I'm seeing three branches in this method:
1. Dynamic PaymentMethod
2. Setting up (PI+sfu, SI)
3. Default (PI)

If that's right, can we 
1. clearly comment these three conditions?
2. (optional, I leave it to you) Rewrite this to match these three branches:
```
if /* dynamic pm */ {
} else if /* setting up */ {
} else {
}
```

Reply:
Yeah I agree the control flow is confusing, but it's because the logic it's modeling is confusing... We are mixing two different payment method type's which leads to weirdness.

We need to know upfront whether the payment method can be represented as an `STPPaymentMethodType`, if it can be we use the requirements as normally defined. However, if it can't be then we need to consider it a dynamic payment method and try to use the requirements defined [here](https://github.com/stripe/stripe-ios/blob/master/StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentMethodType.swift#L18).

However, there is a catch... EVERY payment method can be considered dynamic, since it's just a freeform string [enum case](https://github.com/stripe/stripe-ios/blob/master/StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentMethodType.swift#L40). So we need to know if it is an `STPPaymentMethodType` first, since dynamic is a catch-all.

Take for example Klarna, if we check for a dynamic type first we will end up incorrectly assuming its a dynamic type since it is not [represented](https://github.com/stripe/stripe-ios/blob/master/StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentMethodType.swift#L36) in `PaymentSheet.PaymentMethodType`. So we will create `.dynamic("klarna")`, then we will end up checking the requirements [here](https://github.com/stripe/stripe-ios/blob/master/StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentMethodType.swift#L18) and see there is no Klarna requirements and default to unavailable.

So the order of the control flow matters too. It seemed like a guard was the best way to check for a `STPPaymentMethodType` while not putting everything in-between brackets. I pushed up a slight improvement to the control flow, but it's marginal.